x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,739 @@
+        this.localRepositoryPasswordIdSalt = UUIDs.randomBase64UUID();
+        this.localRepositoryPasswordId = AESKeyUtils.computeId(
+            AESKeyUtils.generatePasswordBasedKey(repositoryPassword, localRepositoryPasswordIdSalt.getBytes(StandardCharsets.UTF_8))
+        );
+        this.validatedRepositoryPasswordId = new AtomicReference<>(this.localRepositoryPasswordId);
+        // stores decrypted DEKs; DEKs are reused to encrypt/decrypt multiple independent blobs
+        this.DEKCache = CacheBuilder.<String, SecretKey>builder().setMaximumWeight(DEK_CACHE_WEIGHT).build();
+        if (isReadOnly() != delegatedRepository.isReadOnly()) {
+            throw new RepositoryException(
+                metadata.name(),
+                "Unexpected fatal internal error",
+                new IllegalStateException("The encrypted repository must be read-only iff the delegate repository is read-only")
+            );
+        }
+    }
+
+    /**
+     * The repository hook method which populates the snapshot metadata with the salted password hash of the repository on the (master)
+     * node that starts of the snapshot operation. All the other actions associated with the same snapshot operation will first verify
+     * that the local repository password checks with the hash from the snapshot metadata.
+     * <p>
+     * In addition, if the installed license does not comply with encrypted snapshots, this throws an exception, which aborts the snapshot
+     * operation.
+     *
+     * See {@link org.elasticsearch.repositories.Repository#adaptUserMetadata(Map)}.
+     *
+     * @param userMetadata the snapshot metadata as received from the calling user
+     * @return the snapshot metadata containing the salted password hash of the node initializing the snapshot
+     */
+    @Override
+    public Map<String, Object> adaptUserMetadata(Map<String, Object> userMetadata) {
+        // because populating the snapshot metadata must be done before the actual snapshot is first initialized,
+        // we take the opportunity to validate the license and abort if non-compliant
+        if (false == licenseStateSupplier.get().isEncryptedSnapshotAllowed()) {
+            throw LicenseUtils.newComplianceException("encrypted snapshots");
+        }
+        Map<String, Object> snapshotUserMetadata = new HashMap<>();
+        if (userMetadata != null) {
+            snapshotUserMetadata.putAll(userMetadata);
+        }
+        // set out the ID of the repository secret
+        // this is then checked before every snapshot operation (i.e. {@link #snapshotShard} and {@link #finalizeSnapshot})
+        // to assure that all participating nodes in the snapshot operation are using the same repository secret
+        snapshotUserMetadata.put(PASSWORD_ID_SALT_USER_METADATA_KEY, localRepositoryPasswordIdSalt);
+        snapshotUserMetadata.put(PASSWORD_ID_USER_METADATA_KEY, localRepositoryPasswordId);
+        logger.trace(
```

Comment:
Okay, I've reverted all uses of the message supplier pattern, as none used parameters that required computation.
I tended to always use a message supplier because I can more easily ignore it if I see it part of every log statement, and it's one less decision to take when coding, but I understand others' reluctance and probably a better strategy is to default on never using the pattern.